File: tutorials/CsTutorials/behaviac/runtime/Base/Properties.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;

namespace behaviac
{
    public class Property
    {
        public Property(CMemberBase pMemberBase, bool bIsConst)
        {
            m_memberBase = pMemberBase;
            m_variableId = 0;
            m_bValidDefaultValue = false;
            m_bIsConst = bIsConst;
        }

        protected Property(Property copy)
        {
            m_variableName = copy.m_variableName;
            m_instanceName = copy.m_instanceName;
            m_variableId = copy.m_variableId;
            m_memberBase = copy.m_memberBase;

            m_parent = copy.m_parent;
            m_index = copy.m_index;

            m_bValidDefaultValue = copy.m_bValidDefaultValue;
            m_defaultValue = copy.m_defaultValue;
            m_bIsConst = copy.m_bIsConst;
            m_bIsStatic = copy.m_bIsStatic;
            m_bIsLocal = copy.m_bIsLocal;
            m_strNativeTypeName = copy.m_strNativeTypeName;
        }

        private IProperty m_bindingProperty;

        public IProperty BindingProperty
        {
            get
            {
                return m_bindingProperty;
            }
            set
            {
                m_bindingProperty = value;
            }
        }

        public string Name
        {
            get
            {
                return this.m_variableName;
            }
            set
            {
                this.m_variableName = value;

                this.m_variableId = Utils.MakeVariableId(this.m_variableName);
            }
        }

        public string InstanceName
        {
            get
            {
                if (!string.IsNullOrEmpty(this.m_instanceName))
                {
                    return this.m_instanceName;
                }

                return m_memberBase != null ? m_memberBase.InstanceName : null;
            }
            set
            {
                this.m_instanceName = value;
            }
        }

        public uint VariableId
        {
            get
            {
                return m_variableId;
            }
        }

        private bool m_bIsStatic;

        public bool IsStatic
        {
            get
            {
                return this.m_bIsStatic;
            }
            set
            {
                this.m_bIsStatic = value;
            }
        }

        private bool m_bIsLocal;

        public bool IsLocal
        {
            get
            {
                return this.m_bIsLocal;
            }
            set
            {
                this.m_bIsLocal = value;
            }
        }

        public Type PropertyType
        {
            get
            {
                if (this.m_defaultValue != null)
                {
                    return this.m_defaultValue.GetType();
                }

                if (this.m_memberBase != null)
                {
                    return this.m_memberBase.MemberType;
                }

                return null;
            }
        }

        private string m_strNativeTypeName;
        public string NativeTypeName
        {
            get
            {
                return m_strNativeTypeName;
            }
            set
            {
                m_strNativeTypeName = Utils.GetNativeTypeName(value).Replace("::", ".");
            }
        }

        public float GetRange()
        {
            if (this.m_memberBase != null)
            {
                return this.m_memberBase.GetRange();
            }

            return 1.0f;
        }

#if K_TYPE_CREATOR_
        template<typename T>
        static bool Register(string typeName)
        {
            PropertyCreators()[typeName] = &Creator<T>;

            return true;
        }
        template<typename T>
        static void UnRegister(string typeName)
        {
            PropertyCreators().erase(typeName);
        }

        static void RegisterBasicTypes();
        static void UnRegisterBasicTypes();
#endif//#if K_TYPE_CREATOR_

        public virtual Property clone()
        {
            Property p = new Property(this);

            return p;
        }

        public static void Cleanup()
        {
        }

        protected string m_variableName;
        protected string m_instanceName;
        protected uint m_variableId;

        protected Property m_parent;
        protected Property m_index;

        protected readonly CMemberBase m_memberBase;

        protected object m_defaultValue;
        private bool m_bValidDefaultValue;
        protected readonly bool m_bIsConst;
    }


}//namespace behaviac

File: src/property/properties.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/common/base.h"
#include "behaviac/property/properties.h"
#include "behaviac/common/workspace.h"
#include "behaviac/agent/agent.h"
#include "behaviac/common/member.h"
#include "behaviac/common/meta.h"

namespace behaviac {
    Variables::Variables() {
        BEHAVIAC_ASSERT(this->m_variables.size() == 0);
    }
    Variables::Variables(const behaviac::map<uint32_t, IInstantiatedVariable*>& vars) {
        this->m_variables = vars;
    }

    Variables::~Variables() {
        this->Clear(true);
    }

    void Variables::Clear(bool bFull) {
        if (bFull) {
            for (Variables_t::iterator it = this->m_variables.begin();
                 it != this->m_variables.end(); ++it) {
                IInstantiatedVariable* pVar = it->second;

                BEHAVIAC_DELETE(pVar);
            }

            this->m_variables.clear();
        } else {
            for (Variables_t::iterator it = this->m_variables.begin();
                 it != this->m_variables.end();) {
                IInstantiatedVariable* pVar = it->second;

                Variables_t::iterator it_temp = it;
                ++it;

                BEHAVIAC_DELETE(pVar);
                this->m_variables.erase(it_temp);

            }
        }
    }

    void Variables::Log(const Agent* pAgent, bool bForce) {
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_UNUSED_VAR(bForce);
#if !BEHAVIAC_RELEASE

        if (Config::IsLoggingOrSocketing()) {
            for (Variables_t::iterator it = this->m_variables.begin();
                 it != this->m_variables.end(); ++it) {
                IInstantiatedVariable* pVar = it->second;

                pVar->Log(pAgent);
            }
        }

#endif//BEHAVIAC_RELEASE
    }

    //void IVariable::CopyTo(Agent* pAgent)
    //{
    //    BEHAVIAC_UNUSED_VAR(pAgent);
    //}

    void Variables::CopyTo(Agent* pAgent, Variables& target) const {
        for (Variables_t::iterator it = target.m_variables.begin();
             it != target.m_variables.end(); ++it) {
            IInstantiatedVariable* pVar = it->second;

            BEHAVIAC_DELETE(pVar);
        }

        target.m_variables.clear();

        for (Variables_t::const_iterator it = this->m_variables.begin();
             it != this->m_variables.end(); ++it) {
            IInstantiatedVariable* pVar = it->second;

            IInstantiatedVariable* pNew = pVar->clone();

            target.m_variables[it->first] = pNew;
        }

        if (pAgent) {
            for (Variables_t::iterator it = target.m_variables.begin();
                 it != target.m_variables.end(); ++it) {
                IInstantiatedVariable* pVar = it->second;
                pVar->CopyTo(pAgent);
            }
        }
    }

    void Variables::Save(IIONode* node) const {
        CIOID  variablesId("vars");
        IIONode* varsNode = node->newNodeChild(variablesId);

        for (Variables_t::const_iterator it = this->m_variables.begin();
             it != this->m_variables.end(); ++it) {
            IInstantiatedVariable* pVar = it->second;

            //skip agent members
            pVar->Save(varsNode);
        }
    }

    void Variables::LoadVars(const behaviac::string& agentTypeStr, IIONode* node, behaviac::map<uint32_t, IInstantiatedVariable*>& vars) {
        CIOID  variablesId("vars");
        IIONode* varsNode = node->findNodeChild(variablesId);

        if (varsNode) {
            CStringCRC agentType(agentTypeStr.c_str());
            AgentMeta* pAgentMeta = AgentMeta::GetMeta(agentType.GetUniqueID());

			if (pAgentMeta) {
				int varsCount = varsNode->getChildCount();

				for (int i = 0; i < varsCount; ++i) {
					IIONode* varNode = varsNode->getChild(i);

					CIOID  nameId("name");
					behaviac::string nameStr;
					varNode->getAttr(nameId, nameStr);

					CIOID  valueId("value");
					behaviac::string valueStr;
					varNode->getAttr(valueId, valueStr);

					CStringCRC memberId(nameStr.c_str());

					IProperty* pProperty = pAgentMeta->GetProperty(memberId.GetUniqueID());

					BEHAVIAC_ASSERT(pProperty);

					if (pProperty) {
						IInstantiatedVariable* p = pProperty->Instantiate();

						vars[memberId.GetUniqueID()] = p;
						p->SetValueFromString(valueStr.c_str());
					}
				}
			}
        }
    }

    void Variables::Load(IIONode* node) {
        CIOID  attrId("agentType");
        behaviac::string agentTypeStr;
        node->getAttr(attrId, agentTypeStr);

        this->LoadVars(agentTypeStr, node, this->m_variables);
    }

    IInstantiatedVariable* Variables::GetVariable(uint32_t varId) const {
        Variables_t::iterator it = ((Variables*)this)->m_variables.find(varId);

        if (it != this->m_variables.end()) {
            return it->second;
        }

        return NULL;
    }

    void Variables::AddVariable(uint32_t varId, IInstantiatedVariable* pVar, int stackIndex) {
		BEHAVIAC_UNUSED_VAR(stackIndex);
		BEHAVIAC_ASSERT(this->m_variables.find(varId) == this->m_variables.end());

        this->m_variables[varId] = pVar;
    }

    void Variables::Unload() {
        for (Variables_t::iterator it = this->m_variables.begin();
             it != this->m_variables.end();) {
            Variables_t::iterator itCurrent = it;
            Variables_t::iterator itNext = ++it;

            IInstantiatedVariable* pVar = itCurrent->second;
            this->m_variables.erase(itCurrent->first);
            it = itNext;
            BEHAVIAC_UNUSED_VAR(pVar);

        }
    }

    void Variables::Unload(const char* variableName) {
        BEHAVIAC_ASSERT(!StringUtils::IsNullOrEmpty(variableName));
        uint32_t varId = MakeVariableId(variableName);

        if (this->m_variables.find(varId) != this->m_variables.end()) {
            this->m_variables.erase(varId);
        }
    }

    void Variables::Cleanup()
    {}
}//namespace behaviac
